src/tls/control.rs

```diff
@@ -81,12 +82,108 @@ impl ControlPlaneClientCertProvider for ControlPlaneAuthentication {
+    /// Will verify the certificate is valid in the following ways:
+    /// - Signed by a  trusted `RootCertStore` CA
+    /// - Not Expired
+    fn verify_server_cert(
+        &self,
+        end_entity: &CertificateDer<'_>,
+        intermediates: &[CertificateDer<'_>],
+        sn: &ServerName,
+        ocsp_response: &[u8],
+        now: UnixTime,
+    ) -> Result<ServerCertVerified, rustls::Error> {
+        let cert = rustls::server::ParsedCertificate::try_from(end_entity)?;
+
+        let algs = provider().signature_verification_algorithms;
+        rustls::client::verify_server_cert_signed_by_trust_anchor(
+            &cert,
+            &self.roots,
+            intermediates,
+            now,
+            algs.all,
+        )?;
+
+        if !ocsp_response.is_empty() {
+            tracing::trace!("Unvalidated OCSP response: {ocsp_response:?}");
+        }
+
+        // First attempt to verify the original server name...
+        if let Err(err) = rustls::client::verify_server_name(&cert, sn) {
+            tracing::debug!(
+                "failed to verify {sn:?} ({err}), attempting alt name {:?}",
+                self.alt_server_name
+            );
+            // That failed, lets try the alternative one
+            rustls::client::verify_server_name(&cert, &self.alt_server_name)?;
+        }
+
+        Ok(ServerCertVerified::assertion())
+    }
+
+    // Rest use the default implementations
+
+    fn verify_tls12_signature(
+        &self,
+        message: &[u8],
+        cert: &CertificateDer<'_>,
+        dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, rustls::Error> {
+        rustls::crypto::verify_tls12_signature(
+            message,
+            cert,
+            dss,
+            &provider().signature_verification_algorithms,
+        )
+    }
+
+    fn verify_tls13_signature(
+        &self,
+        message: &[u8],
+        cert: &CertificateDer<'_>,
+        dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, rustls::Error> {
+        rustls::crypto::verify_tls13_signature(
+            message,
+            cert,
+            dss,
+            &provider().signature_verification_algorithms,
+        )
+    }
+
+    fn supported_verify_schemes(&self) -> Vec<SignatureScheme> {
+        provider()
+            .signature_verification_algorithms
+            .supported_schemes()
+    }
+}
+
+async fn control_plane_client_config(
+    root_cert: &RootCert,
+    alt_hostname: Option<String>,
+) -> Result<ClientConfig, Error> {
     let roots = root_to_store(root_cert).await?;
-    Ok(ClientConfig::builder_with_provider(provider())
-        .with_protocol_versions(crate::tls::TLS_VERSIONS)?
-        .with_root_certificates(roots)
-        .with_no_client_auth())
+    let c = ClientConfig::builder_with_provider(provider())
+        .with_protocol_versions(crate::tls::TLS_VERSIONS)?;
+    if let Some(alt_hostname) = alt_hostname {
```

Comment:
An `info` would be nice here, to make it clear that because you set an alt_hostname, we're using "nonstandard" cert validation logic (ATM it's copypasted from rusttls mostly, sure, but good to flag).